zetaclient/evm/evm_rpc_fallback.go

```diff
@@ -0,0 +1,398 @@
+	var res []ethtypes.Log
+	var err error
+
+	for i := 0; i < e.ethClients.Length(); i++ {
+		if client := e.ethClients.First(); client != nil {
+			res, err = client.(interfaces.EVMRPCClient).FilterLogs(ctx, query)
+		}
+		if err != nil {
+			e.logger.Debug().Err(err).Msg("client endpoint failed attempting fallback client")
+			e.ethClients.Next()
+			continue
+		}
+		break
+	}
+	return res, err
+}
+
+func (e *EthClientFallback) SubscribeFilterLogs(ctx context.Context, query ethereum.FilterQuery, ch chan<- ethtypes.Log) (ethereum.Subscription, error) {
+	var res ethereum.Subscription
+	var err error
+
+	for i := 0; i < e.ethClients.Length(); i++ {
+		if client := e.ethClients.First(); client != nil {
+			res, err = client.(interfaces.EVMRPCClient).SubscribeFilterLogs(ctx, query, ch)
+		}
+		if err != nil {
+			e.logger.Debug().Err(err).Msg("client endpoint failed attempting fallback client")
+			e.ethClients.Next()
+			continue
+		}
+		break
+	}
+	return res, err
+}
+
+func (e *EthClientFallback) SendTransaction(ctx context.Context, tx *ethtypes.Transaction) error {
+	var err error
+
+	for i := 0; i < e.ethClients.Length(); i++ {
+		if client := e.ethClients.First(); client != nil {
+			err = client.(interfaces.EVMRPCClient).SendTransaction(ctx, tx)
+		}
+		if err != nil {
+			e.logger.Debug().Err(err).Msg("client endpoint failed attempting fallback client")
+			e.ethClients.Next()
+			continue
+		}
+		break
+	}
+	return err
+}
+
+func (e *EthClientFallback) BlockNumber(ctx context.Context) (uint64, error) {
+	var res uint64
+	var err error
+
+	for i := 0; i < e.ethClients.Length(); i++ {
+		if client := e.ethClients.First(); client != nil {
+			res, err = client.(interfaces.EVMRPCClient).BlockNumber(ctx)
+		}
+		if err != nil {
+			e.logger.Debug().Err(err).Msg("client endpoint failed attempting fallback client")
+			e.ethClients.Next()
+			continue
+		}
+		break
+	}
+	return res, err
+}
+
+func (e *EthClientFallback) BlockByNumber(ctx context.Context, number *big.Int) (*ethtypes.Block, error) {
+	var res *ethtypes.Block
+	var err error
+
+	for i := 0; i < e.ethClients.Length(); i++ {
+		if client := e.ethClients.First(); client != nil {
+			res, err = client.(interfaces.EVMRPCClient).BlockByNumber(ctx, number)
+		}
+		if err != nil {
+			e.logger.Debug().Err(err).Msg("client endpoint failed attempting fallback client")
+			e.ethClients.Next()
+			continue
+		}
+		break
+	}
+	return res, err
+}
+
+func (e *EthClientFallback) TransactionByHash(ctx context.Context, hash ethcommon.Hash) (tx *ethtypes.Transaction, isPending bool, err error) {
+	for i := 0; i < e.ethClients.Length(); i++ {
+		if client := e.ethClients.First(); client != nil {
+			tx, isPending, err = client.(interfaces.EVMRPCClient).TransactionByHash(ctx, hash)
+		}
+		if err != nil {
+			e.logger.Debug().Err(err).Msg("client endpoint failed attempting fallback client")
+			e.ethClients.Next()
```

Comment:
Seems the rotation happens on every error of the `First` endpoint. It gives me a feeling that the underlying endpoint rotation can happen often and sometimes unexpectly across invocations of this `Next` method. The reason is that some errors from RPC call are normal and expected, for example:
- `TransactionByHash` can return 'Not Found' when non-existing `hash` is provided (can be a outbound tx in mempool waiting for inclusion).
- `TransactionReceipt` can return 'Not Found' when tx is waiting for inclusion.
- `TransactionSender` usually return errors on Goerli testnet.
- `SendTransaction` can return error on nonce mismatch, sending duplicate tx, etc.
It's hard to enumerate all RPC errors across RPC providers and to tell which ones should trigger rotation. Is it a good idea to reuse and repurpose the RPC status monitoring thread `WatchRPCStatus` for RPC rotation goal. 

There can be different options regarding impl. A quick impl in my mind looks like:
- `WatchRPCStatus` does the basic check as it is doing now in the thread.
-  Reduce the ticker of RPC status check a bit to a shorter period, e.g., `60 secs` --> `15 secs`
-  Define a `RPCRotationFailRateThrehold` to indicate the max percentage of failed RPC status checks in a health check time window. e.g., 30%. Also, define a `RPCHealthCheckWindow` to, e.g., 10 tickers.
-  If the status check fail rate exceeds the threhold (g.g., > 30%) in the latest time window, zetaclient then trigger RPC rotation.